src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+            assert(table.content_blocks_received.count() <= table.table_blocks_total.? - 1);
+
+            table.table_blocks_written += 1;
+            if (table.table_blocks_written == table.table_blocks_total.?) {
+                queue.faulty_tables.remove(table);
+                (table.callback)(table, .repaired);
+            }
+        }
+    }
+
+    fn enqueue_table_content(
+        queue: *GridRepairQueue,
+        table: *RepairTable,
+        index_block_data: BlockPtrConst,
+    ) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_blocks.count() ==
+            queue.enqueued_single_blocks + queue.enqueued_table_blocks);
+        assert(table.table_blocks_total == null);
+        assert(table.table_blocks_written == 0);
+        assert(table.content_blocks_received.count() == 0);
+
+        const index_schema = schema.TableIndex.from(index_block_data);
+        const index_block_header = schema.header_from_block(index_block_data);
+        assert(index_block_header.op == table.index_address);
+        assert(index_block_header.checksum == table.index_checksum);
+        assert(schema.BlockType.from(index_block_header.operation) == .index);
+
+        const content_blocks_total = index_schema.content_blocks_used(index_block_data);
+        table.table_blocks_total = 1 + content_blocks_total;
+
+        var content_block_index: u32 = 0;
+        while (content_block_index < content_blocks_total) : (content_block_index += 1) {
+            const block_id = index_schema.content_block(index_block_data, content_block_index);
+
+            const enqueue = queue.enqueue_faulty_block(
+                block_id.block_address,
+                block_id.block_checksum,
+                .{ .table_content = .{ .table = table, .index = content_block_index } },
+            );
+
+            if (enqueue == .replace) {
+                if (enqueue.replace.state == .writing) {
+                    table.content_blocks_received.set(content_block_index);
+                }
+            } else {
+                assert(enqueue == .insert);
+            }
+        }
+    }
+
+    fn release_fault(queue: *GridRepairQueue, fault_index: usize) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_blocks_repair_index < queue.faulty_blocks.count());
+
+        switch (queue.faulty_blocks.entries.items(.value)[fault_index].progress) {
+            .block => queue.enqueued_single_blocks -= 1,
+            .table_index => queue.enqueued_table_blocks -= 1,
+            .table_content => queue.enqueued_table_blocks -= 1,
+        }
+
+        queue.faulty_blocks.swapRemoveAt(fault_index);
+
+        if (queue.faulty_blocks_repair_index == queue.faulty_blocks.count()) {
+            queue.faulty_blocks_repair_index = 0;
+        } else {
+            if (queue.faulty_blocks_repair_index > fault_index) {
+                queue.faulty_blocks_repair_index -= 1;
+            }
```

Comment:
Why do we need this `-= 1`? We are swap-removing, so we don't need to shift the index outside of `== .count()` case handledd above?

I think, as written, this will cause us to send the same block in two consequitive requests?

Reply:
Ah, you're right, this is unnecessary -- I wasn't considering the swap-remove!